README.md
# Handy-Types

![Module Type](https://img.shields.io/badge/Module%20Type-UMD-brightgreen)
![Npm Version](https://img.shields.io/npm/v/handy-types)
![GitHub Tag](https://img.shields.io/github/v/tag/h-sifat/handy-types)
![GitHub Issues](https://img.shields.io/github/issues/h-sifat/handy-types)

A simple and lite weight validation library to reduce boilerplate in your
JavaScript / TypeScript validation code.

It consists of many utility type predicate functions like `"integer"`,
`"positive_integer"`, `"non_empty_string"`, `"plain_object"` and so on. Still
not impressed? How about `"string[]"` to represent a string array or
`"non_empty_string | non_empty_string[]"` to represent a non empty string or an
array of non empty string(s)? TypeScript type annotation is also supported. The
library is fully tested and has **100%** test coverage.

## Importing

The library uses **UMD** module system so it's compatible with all JavaScript
module systems.

### JavaScript / TypeScript

```js
// es modules
import { is, assert, cache, handyTypes, typeNames } from "handy-types";

// commonjs modules
const { is, assert, cache, handyTypes, typeNames } = require("handy-types");
```

### In HTML with the `<script>` tag

```html
<script src="https://unpkg.com/handy-types"></script>
<!-- it will be available as a global object by the name "handy_types" -->
```

## Usages

This library exposes the following 5 entities

1. **handyTypes:** An object containing all the type predicate functions. For
   example, `handyTypes.positive_integer(2); // true`.

1. **typeNames:** An object containing all the names of predicate functions of
   `handyTypes`.
   For example: `typeNames["positive_integer"]; // "Positive Integer"`.
   It may be used to generate meaningful error messages.

1. **is:** a predicate function that uses all the functions in `handyTypes`
   object to validate data. For example, `is("string", "a string"); // true`. It
   has a `cache` method with the same function signature as itself. The
   `it.cache()` method can be used to cache parsed schemas to improve
   performance.

1. **assert:** a utility function similar to `is` but used for making
   assertions. It also has a `cache` method similar to `is`.

1. **cache:** An object used to manage schema caches of `is.cache()` and
   `assert.cache()` functions.

### Type schema syntax

A type schema is a string passed into the `is` and `assert` function to
represent a type. There are three types of type schema:

1. **Basic:** Just a simple handy type name such as `"string"`,
   `"non_empty_array"` etc.

1. **Array:** If we add the `"[]"` suffix after any handy type name it
   represents an array of that type. So `"string[]"` would represent and array
   of string.

1. **Union:** We can combine two or more type schema with a pipe `"|"` character
   to represent an union. For example, `"string | string[]"` to represent a
   string or an array of string(s).

## Usages of `is()`

The `is` predicate function has the following signature.

```ts
interface Is {
  <Type>(schema: string, value: unknown): value is Type;
  cache<Type>(schema: string, value: unknown): value is Type;
}
```

The reason it's a generic function with a type parameter named `Type` is to
support typescript type annotation. For example:

#### TypeScript Example

```ts
let value: unknown;

if (is<string | string[]>("non_empty_string | non_empty_string[]", value)) {
  value; // let value: string | string[]
}
```

In the if block the type of `value` variable is `string | string[]`. We've to
pass the type of value (`string | string[]`) manually because it's not possible
to process an union schema with TypeScript to determine it's actual type.

But if this seems a little bit of extra work to you then you can use the basic
type predicate functions directly from the `handyTypes` object . For
example:

```ts
let value: unknown;

if (handyTypes.integer(value)) {
  value; // let value: number
}
```

Here in the if block the type of `value` variable will be set to
`number` automatically. But the downsides of this approach are:

1. We can't use array or union type schemas
2. `handyTypes.integer(value)` doesn't seem intuitive because most of the time
   a predicate function starts with the word **is** as a convention.

#### JavaScript Example

For JavaScript just remove the generic type argument.

```js
const hobbies = "programming";
if (is("non_empty_string | non_empty_string[]", hobbies)) {
  hobbies;
  // so `hobbies` is either a non_empty_string or a
  // non_empty_string array
}
```

### Usages of `is.cache()`

Use the `is.cache()` function instead of `is` for **array** and **union**
schemas to improve performance. It will parse and cache the schema so that it
doesn't have to waste time parsing the same schema again and again.

#### TypeScript Example

```ts
if (is.cache<string | string[]>("string | string[]", value)) {
  value; // here value is of type: string | string[]
}
```

#### JavaScript Example

```ts
if (is.cache("string | string[]", value)) {
  value; // here value is of type: string | string[]
}
```

## Usages of `assert`

We can use the `assert` function to make assertions. It has the following
function signature:

```ts
interface Assert {
  <Type>(
    schema: string,
    value: unknown,
    errorInfo?: ErrorInformation
  ): asserts value is Type;
  cache<Type>(
    schema: string,
    value: unknown,
    errorInfo?: ErrorInformation
  ): asserts value is Type;
}
```

Here `ErrorInformation` refers to the interface below.

```ts
interface ErrorInformation {
  name?: string;
  message?: string;
  code?: string | number;
  otherInfo?: object;
}
```

Just like the `is` function it takes a type schema and the variable we're making
assertion on as it's first and second arguments respectively. Then we can
provide more information in the `errorInfo` object to customize the error
object.

Examples:

```ts
let value: unknown;

assert<number>("integer", value);
// throws error: `Value must be of type Integer`

assert<number>("integer", value, {
  name: "Age",
  code: "INVALID_AGE",
});
// throws error: `Age must be of type Integer`, with code: "INVALID_AGE"

// Use custom message instead of generating one
assert<string>("non_empty_string", value, {
  message: "Invalid path",
  otherInfo: {
    path: value,
    errorId: -3,
  },
});
// throws error: `Invalid path` , path: undefined, errorId: -3
```

### Usages of `assert.cache()`

It serves the same purpose as `is.cache()`, it caches parsed schemas.

```ts
assert.cache<string | string[]>(
  "non_empty_string | non_empty_string[]",
  value,
  { name: "hobbies" }
); // use caching for improved performance
```

## Usages of the `cache` object

The `cache` object can be used to manage schema caches. It has the following
interface.

```ts
Readonly<{
  readonly size: number;
  has(schema: string): boolean;
  delete(schema: string): boolean;
  clear(): void;
}>;
```

Examples:

```ts
console.log(cache.size); // 0

const schema = "integer | integer[]";

is.cache<number | number[]>(schema, 23);

console.log(cache.size); // 1

console.log(cache.has(schema)); // true

// use the delete method to delete a specific schema cache
cache.delete(schema); // true

console.log(cache.size); // 0

// clear all caches
cache.clear();
```

## All handy types

Below are the lists of all the type predicates available in the `handyTypes`
object.

### Base Types

| Type Name   | Full Name   | Implementation                                      |
| ----------- | ----------- | --------------------------------------------------- |
| boolean     | Boolean     | `typeof value === "boolean"`                        |
| symbol      | Symbol      | `typeof value === "symbol"`                         |
| string      | String      | `typeof value === "string"`                         |
| object      | Object      | `typeof value === "object"`                         |
| big_integer | Big Integer | `typeof value === "bigint"`                         |
| function    | Function    | `typeof value === "function"`                       |
| undefined   | Undefined   | `typeof value === "undefined"`                      |
| **number**  | Number      | `typeof value === "number" && !Number.isNaN(value)` |

**Note:** The type `number` is not just `typeof value === "number"`!

### Number Types

| Type Name           | Full Name           | Implementation                   |
| ------------------- | ------------------- | -------------------------------- |
| finite_number       | Finite Number       | `Number.isFinite(n)`             |
| positive_number     | Positive Number     | `handyTypes.number(n) && n > 0`  |
| non_negative_number | Non Negative Number | `handyTypes.number(n) && n >= 0` |
| negative_number     | Negative Number     | `handyTypes.number(n) && n < 0`  |
| non_positive_number | Non Positive Number | `handyTypes.number(n) && n <= 0` |

### Integer Types

| Type Name            | Full Name            | Implementation                  |
| -------------------- | -------------------- | ------------------------------- |
| integer              | Integer              | `Number.isInteger(i)`           |
| safe_integer         | Safe Integer         | `Number.isSafeInteger(i)`       |
| positive_integer     | Positive Integer     | `Number.isInteger(i) && i > 0`  |
| non_negative_integer | Non Negative Integer | `Number.isInteger(i) && i >= 0` |
| negative_integer     | Negative Integer     | `Number.isInteger(i) && i < 0`  |
| non_positive_integer | Non Positive Integer | `Number.isInteger(i) && i <= 0` |

### Bytewise Integer Types

| Type Name              | Full Name               | Range                                   |
| ---------------------- | ----------------------- | --------------------------------------- |
| 8bit_integer           | 8 Bit Integer           | **-128** to **127**                     |
| 8bit_unsigned_integer  | 8 Bit Unsigned Integer  | **0** to **255**                        |
| 16bit_integer          | 16 Bit Integer          | **-32,768** to **32,767**               |
| 16bit_unsigned_integer | 16 Bit Unsigned Integer | **0** to **65,535**                     |
| 32bit_integer          | 32 Bit Integer          | **-2,147,483,648** to **2,147,483,647** |
| 32bit_unsigned_integer | 32 Bit Unsigned Integer | **0** to **4,294,967,295**              |

### Array Types

| Type Name       | Full Name       | Implementation                               |
| --------------- | --------------- | -------------------------------------------- |
| array           | Array           | `Array.isArray(value)`                       |
| non_empty_array | Non-Empty Array | `Array.isArray(value) && value.length !== 0` |

### Object Types

| Type Name       | Full Name       | Implementation                                                         |
| --------------- | --------------- | ---------------------------------------------------------------------- |
| non_null_object | Non-Null Object | `typeof value === "object" && value !== null`                          |
| plain_object    | Plain Object    | `typeof value === "object" && value !== null && !Array.isArray(value)` |

### String Types

| Type Name                | Full Name                  | Implementation                                       |
| ------------------------ | -------------------------- | ---------------------------------------------------- |
| non_empty_string         | Non-Empty String           | `typeof value === "string" && value !== ""`          |
| trimmed_non_empty_string | Non-Empty String (trimmed) | `typeof value === "string" && !!value.trim().length` |

### Other Types

| Type Name   | Full Name    | Implementation                            |
| ----------- | ------------ | ----------------------------------------- |
| nan         | Not A Number | `Number.isNaN(value)`                     |
| any         | Any          | `true` // returns true for any value      |
| nullish     | Nullish      | `value === null \|\| value === undefined` |
| non_nullish | Non-Nullish  | `value !== null && value !== undefined`   |

If you find any bug or want to improve something please feel free to open an
issue. Pull requests are also welcomed.


CHANGELOG.md
# Changelog

## v2.1.0

- Add `trimmed_non_empty_string` type

## v2.0.0

- Remove unnecessary handy types such as "empty_string", "empty_array" and so on
- Use standard mathematical definitions for number types
- Add `is` predicate function for extended validation
- Add `assert` function for making assertions

## v1.4.0

- Added more number types

## v1.3.0

- Added "nullish" type
- Fixed type annotation for "empty_string"

## v1.2.0

- Added alias "num"
- Added clarification for "even" and "odd" types

## v1.1.0

- Added types **int8**, **uint8**, **int16**, **uint16**, **int32**, and **uint32**
- Added type alias **integer**

## v1.0.0

- Initial release


tests/is.test.ts
import is from "../src/is";

describe("Validation", () => {
  it.concurrent.each([
    {
      type: "string",
      value: 23,
      result: false,
    },
    {
      type: "string ", // with a whitespace at the end
      value: 23,
      result: false,
    },
    {
      type: "string",
      value: "a string",
      result: true,
    },
    {
      type: "  string[]   ",
      value: "not a string array",
      result: false,
    },
    {
      type: "string[]",
      value: ["a", "b"],
      result: true,
    },
    {
      type: "string | number",
      value: "a_string",
      result: true,
    },
    {
      type: "string | number",
      value: 231,
      result: true,
    },
    {
      type: "string | number[]",
      value: [231, 2],
      result: true,
    },
  ])(`is("$type", $value) === $result`, ({ type, value, result }) => {
    expect(is(type, value)).toBe(result);
  });
});

describe("Error Handling", () => {
  it.concurrent.each([
    23,
    { val: "non a string" },
    "not_valid_handy_type",
    "not_valid_handy_type[]",
    "not_valid_handy_type[] | integer",
  ])("throws error if schema (%p) is invalid", (schema) => {
    expect.assertions(1);
    try {
      // @ts-ignore
      is(schema, "value");
    } catch (ex: any) {
      expect(ex.code).toBe("INVALID_SCHEMA");
    }
  });
});


tests/util.test.ts
import parseSchema from "../src/schema-parser";
import { assertValidHandyType, EPP, makeCacheStore } from "../src/util";

describe("makeCacheStore", () => {
  const { publicInterface, privateInterface } = makeCacheStore();

  beforeEach(() => {
    publicInterface.clear();
  });

  const schema = "integer[]";
  const schemaDefinition = parseSchema(schema);

  describe("privateInterface", () => {
    it("set(): stores a schema", () => {
      privateInterface.set(schema, schemaDefinition);
      expect(privateInterface.has(schema)).toBeTruthy();
      expect(privateInterface.get(schema)).toEqual(schemaDefinition);
    });
  });

  describe("publicInterface", () => {
    it("delete(): deletes a schema", () => {
      expect(publicInterface.size).toBe(0);

      privateInterface.set(schema, schemaDefinition);

      expect(publicInterface.size).toBe(1);
      expect(publicInterface.has(schema)).toBeTruthy();

      expect(publicInterface.delete(schema)).toBeTruthy();

      expect(publicInterface.size).toBe(0);
      expect(publicInterface.has(schema)).toBeFalsy();
    });

    it("clear(): clears all schema", () => {
      privateInterface.set(schema, schemaDefinition);
      privateInterface.set(schema + "_", schemaDefinition);

      expect(publicInterface.size).toBe(2);
      publicInterface.clear();
      expect(publicInterface.size).toBe(0);
    });
  });
});

describe("EPP", () => {
  it("adds an error code in the error object", () => {
    const message = "oops!";
    const code = "ERR_CODE";
    const error = new EPP(message, code);

    expect(error).toHaveProperty("message", message);
    expect(error).toHaveProperty("code", code);
  });
});

describe("assertValidHandyType", () => {
  it("throws error if the given schema is not a handy type", () => {
    expect.assertions(1);
    try {
      assertValidHandyType("duck");
    } catch (ex: any) {
      expect(ex.code).toBe("INVALID_SCHEMA");
    }
  });

  it("does not throw an error if the given schema is a valid handy type", () => {
    expect(() => {
      assertValidHandyType("integer");
    }).not.toThrow(Error);
  });
});


tests/index.test.ts
import { is, assert, cache } from "../src";

beforeEach(() => {
  cache.clear();
});

describe("is", () => {
  it("validates the input", () => {
    expect(is("integer | integer[]", 23)).toBeTruthy();
    expect(is.cache("integer | integer[]", 23)).toBeTruthy();
    expect(is("integer | integer[]", ["not_a_integer"])).toBeFalsy();
    expect(is.cache("integer | integer[]", ["not_a_integer"])).toBeFalsy();
  });

  it("uses caching", () => {
    expect(cache.size).toBe(0);

    const schema = "integer | integer[]";
    expect(is.cache(schema, [23])).toBeTruthy();

    expect(cache.size).toBe(1);

    // remove cache
    cache.delete(schema);
    expect(cache.size).toBe(0);
  });
});

describe("assert", () => {
  it("validates the input", () => {
    expect(() => {
      assert("integer | integer[]", [23]);
    }).not.toThrow(Error);

    expect(() => {
      assert.cache("integer | integer[]", [23]);
    }).not.toThrow(Error);

    expect(() => {
      assert("integer | integer[]", "a string");
    }).toThrow(/Integer/);

    expect(() => {
      assert.cache("integer | integer[]", "a string");
    }).toThrow(/Integer/);
  });

  it("uses caching", () => {
    expect(cache.size).toBe(0);

    const schema = "integer | integer[]";
    expect(() => {
      assert.cache(schema, [23]);
    }).not.toThrow();

    expect(cache.size).toBe(1);
    expect(cache.has(schema)).toBeTruthy();

    // remove cache
    cache.clear();
    expect(cache.size).toBe(0);
  });
});


src/is.ts
import { handyTypes, AllHandyTypes } from "./types";
import { EPP, assertValidHandyType } from "./util";

export default function is<Type>(schema: string, value: any): value is Type {
  if (!handyTypes.non_empty_string(schema))
    throw new EPP(`Invalid schema: ${schema}`, "INVALID_SCHEMA");

  schema = schema.trim();
  // if the schema is already a handy-type then we don't have to check
  // whether it's a union or array schema.
  if (schema in handyTypes)
    return validateBasicType({ type: schema as AllHandyTypes, value });

  // it's a union schema
  if (schema.includes("|")) {
    const subSchemas = schema.split("|").map((subSchema) => subSchema.trim());

    return validateUnionType({ subSchemas, value });
  }

  // it's an array schema
  if (schema.endsWith("[]")) {
    const elementType = schema.slice(0, -2); // remove "[]"
    assertValidHandyType(elementType);

    return validateArrayType({
      array: value,
      elementType: elementType as AllHandyTypes,
    });
  }

  throw new EPP(`Invalid handy-type: "${schema}".`, "INVALID_SCHEMA");
}

// -------------- validateBasicType ---------------------------------
interface ValidateBasicTypeArgument {
  type: AllHandyTypes;
  value: unknown;
}

export function validateBasicType(arg: ValidateBasicTypeArgument): boolean {
  const { type, value } = arg;
  return handyTypes[type](value);
}

// -------------- validateUnionType ---------------------------------
interface validateUnionTypeArgument {
  subSchemas: string[];
  value: unknown;
}

function validateUnionType(arg: validateUnionTypeArgument): boolean {
  const { subSchemas, value } = arg;

  for (let index = 0; index < subSchemas.length; index++) {
    const subSchema = subSchemas[index];

    try {
      if (is(subSchema, value)) return true;
    } catch (ex: any) {
      const message = `Invalid sub schema ("${subSchema}") at position ${
        index + 1
      } in union schema "${subSchema}".`;

      throw new EPP(message, "INVALID_SCHEMA");
    }
  }

  return false;
}
// -------------- validateArrayType ---------------------------------
type ValidateArrayTypeArgument = {
  array: unknown;
  elementType: AllHandyTypes;
};

export function validateArrayType(arg: ValidateArrayTypeArgument): boolean {
  const { array, elementType } = arg;

  if (!handyTypes.array(array)) return false;

  const isAnyElementInvalid = array.some(
    (element) => !handyTypes[elementType](element)
  );

  return !isAnyElementInvalid;
}


src/util.ts
import type { SchemaDefinition } from "./schema-parser";
import { handyTypes, AllHandyTypes } from "./types";

export class EPP extends Error {
  constructor(message: string, public code: string) {
    super(message);
  }
}

export function assertValidHandyType(
  type: string
): asserts type is AllHandyTypes {
  if (!(type in handyTypes))
    throw new EPP(`Invalid handy type: "${type}"`, "INVALID_SCHEMA");
}

export function makeCacheStore() {
  const cacheStore: Map<string, SchemaDefinition> = new Map();

  const publicInterface = Object.freeze({
    get size() {
      return cacheStore.size;
    },
    has(schema: string) {
      return cacheStore.has(schema);
    },
    delete(schema: string) {
      return cacheStore.delete(schema);
    },
    clear() {
      cacheStore.clear();
    },
  });

  const privateInterface = Object.freeze({
    has(schema: string) {
      return cacheStore.has(schema);
    },
    get(schema: string) {
      return cacheStore.get(schema);
    },
    set(schema: string, schemaDefinition: SchemaDefinition): void {
      cacheStore.set(schema, schemaDefinition);
    },
  });

  return { privateInterface, publicInterface };
}


src/index.ts
import _is from "./is";
import { makeCacheStore } from "./util";
import type { ErrorInformation } from "./assert";
import schemaWiseIs from "./schemawise-validator";
import _assert, { schemaWiseAssert } from "./assert";
import parseSchema, { SchemaDefinition } from "./schema-parser";
import { handyTypes, typeNames, allTypeCategories } from "./types";

interface Is {
  <Type>(schema: string, value: unknown): value is Type;
  cache<Type>(schema: string, value: unknown): value is Type;
}

interface Assert {
  <Type>(
    schema: string,
    value: unknown,
    errorInfo?: ErrorInformation
  ): asserts value is Type;
  cache<Type>(
    schema: string,
    value: unknown,
    errorInfo?: ErrorInformation
  ): asserts value is Type;
}

const {
  publicInterface: publicCacheInterface,
  privateInterface: privateCacheInterface,
} = makeCacheStore();

// --------------------- is --------------------
let is: Is;
// @ts-ignore
is = _is;
is.cache = <Type>(schema: string, value: unknown): value is Type => {
  const schemaDefinition = getSchemaDefinition({
    schema,
    privateCacheInterface,
  });

  return schemaWiseIs({ value, schema: schemaDefinition });
};

Object.freeze(is);
// --------------------- assert --------------------
let assert: Assert;
// @ts-ignore
assert = _assert;
assert.cache = <Type>(
  schema: string,
  value: unknown,
  errorInfo?: ErrorInformation
): asserts value is Type => {
  const schemaDefinition = getSchemaDefinition({
    schema,
    privateCacheInterface,
  });
  schemaWiseAssert(schemaDefinition, value, errorInfo);
};

Object.freeze(assert);

// -------------- exports --------------------------

export {
  is,
  assert,
  typeNames,
  handyTypes,
  allTypeCategories,
  publicCacheInterface as cache,
};

// -------------- utility --------------------------
interface GetSchemaDefinitionArgument {
  schema: string;
  privateCacheInterface: typeof privateCacheInterface;
}

function getSchemaDefinition(
  arg: GetSchemaDefinitionArgument
): SchemaDefinition {
  const { schema, privateCacheInterface } = arg;

  if (privateCacheInterface.has(schema))
    return privateCacheInterface.get(schema)!;
  else {
    const schemaDefinition = parseSchema(schema);
    privateCacheInterface.set(schema, schemaDefinition);
    return schemaDefinition;
  }
}


src/assert.ts
import parseSchema, { SchemaDefinition } from "./schema-parser";
import validate from "./schemawise-validator";
import { handyTypes } from "./types";

export interface ErrorInformation {
  name?: string;
  message?: string;
  code?: string | number;
  otherInfo?: object;
}

export default function assert<Type>(
  schema: string,
  value: unknown,
  errorInfo: ErrorInformation = {}
): asserts value is Type {
  const schemaDefinition = parseSchema(schema);
  schemaWiseAssert(schemaDefinition, value, errorInfo);
}

export function schemaWiseAssert<Type>(
  schemaDefinition: SchemaDefinition,
  value: unknown,
  errorInfo: ErrorInformation = {}
): asserts value is Type {
  const isValid = validate({ schema: schemaDefinition, value });
  if (!isValid) throw generateError({ schemaDefinition, errorInfo });
}

interface GenerateErrorArgument {
  schemaDefinition: SchemaDefinition;
  errorInfo?: ErrorInformation;
}

export function generateError(arg: GenerateErrorArgument): Error {
  const { schemaDefinition, errorInfo = {} } = arg;

  const errorMessage =
    "message" in errorInfo
      ? (errorInfo.message as string)
      : generateErrorMessage({
          valueName: errorInfo.name!,
          schemaName: schemaDefinition.schemaName,
        });

  const error = new Error(errorMessage);

  const { code, otherInfo } = errorInfo;

  // @ts-ignore
  if (code) error.code = code;

  if (otherInfo && handyTypes.plain_object(otherInfo)) {
    const { message, code, ...rest } = otherInfo as any;
    Object.assign(error, rest);
  }

  return error;
}

interface GenerateErrorMessageArgument {
  schemaName: string;
  valueName?: string;
}
export function generateErrorMessage(
  arg: GenerateErrorMessageArgument
): string {
  const { schemaName, valueName } = arg;
  return `${valueName || "Value"} must be of type: ${schemaName}`;
}


src/types.ts
type AllBasicTypes =
  | boolean
  | number
  | string
  | symbol
  | bigint
  | object
  | undefined
  | null // I know that null is not a type
  | Function;

export const allTypeCategories = Object.freeze([
  "undefined",
  "boolean",
  "number",
  "string",
  "symbol",
  "object",
  "array",
  "function",
  "other",
  "bigint",
] as const);

type TypePredicate<
  Type,
  Category extends (typeof allTypeCategories)[number]
> = {
  (v: unknown): v is Type;
  category: Category;
};

type NumberPredicate = TypePredicate<number, "number">;
type StringPredicate = TypePredicate<string, "string">;
type ObjectPredicate = TypePredicate<object, "object">;
type ArrayPredicate = TypePredicate<unknown[], "array">;
type NullableObjectPredicate = TypePredicate<object | null, "object">;

type NullishValue = undefined | null;

export interface HandyTypes_Interface {
  // Base types
  undefined: TypePredicate<undefined, "undefined">;
  boolean: TypePredicate<boolean, "boolean">;
  number: NumberPredicate;
  big_integer: TypePredicate<BigInt, "bigint">;
  symbol: TypePredicate<Symbol, "symbol">;
  string: StringPredicate;
  object: NullableObjectPredicate;
  function: TypePredicate<Function, "function">;

  // Number types
  finite_number: NumberPredicate;

  positive_number: NumberPredicate;
  non_negative_number: NumberPredicate;

  negative_number: NumberPredicate;
  non_positive_number: NumberPredicate;

  // Integer Types
  integer: NumberPredicate;
  safe_integer: NumberPredicate;

  positive_integer: NumberPredicate;
  non_negative_integer: NumberPredicate;

  negative_integer: NumberPredicate;
  non_positive_integer: NumberPredicate;

  // Byte wise int
  "8bit_integer": NumberPredicate;
  "8bit_unsigned_integer": NumberPredicate;
  "16bit_integer": NumberPredicate;
  "16bit_unsigned_integer": NumberPredicate;
  "32bit_integer": NumberPredicate;
  "32bit_unsigned_integer": NumberPredicate;

  // String types
  non_empty_string: StringPredicate;
  trimmed_non_empty_string: StringPredicate;

  // Array Types
  array: ArrayPredicate;
  non_empty_array: ArrayPredicate;

  // Object Types
  plain_object: ObjectPredicate;
  non_null_object: ObjectPredicate;

  // Other types
  any: TypePredicate<any, "other">;
  nan: TypePredicate<typeof NaN, "other">;
  nullish: TypePredicate<NullishValue, "other">;
  non_nullish: {
    (v: unknown): v is Exclude<AllBasicTypes, NullishValue>;
    category: "other";
  };
}

const handyTypes: Partial<HandyTypes_Interface> = {};

// Implementation -------------------------------------------------------------
handyTypes.undefined = Object.assign(
  (v: unknown): v is undefined => typeof v === "undefined",
  { category: "undefined" } as const
);

handyTypes.number = Object.assign(
  (v: unknown): v is number => typeof v === "number" && !Number.isNaN(v),
  { category: "number" } as const
);
handyTypes.positive_number = Object.assign(
  (v: unknown): v is number => handyTypes.number!(v) && v > 0,
  { category: "number" } as const
);

handyTypes.non_negative_number = Object.assign(
  (v: unknown): v is number => handyTypes.number!(v) && v >= 0,
  { category: "number" } as const
);

handyTypes.negative_number = Object.assign(
  (v: unknown): v is number => handyTypes.number!(v) && v < 0,
  { category: "number" } as const
);

handyTypes.non_positive_number = Object.assign(
  (v: unknown): v is number => handyTypes.number!(v) && v <= 0,
  { category: "number" } as const
);

handyTypes.finite_number = Object.assign(
  (v: unknown): v is number => Number.isFinite(v),
  { category: "number" } as const
);

// Integer Types ===============
handyTypes.integer = Object.assign(
  (v: unknown): v is number => Number.isInteger(v),
  { category: "number" } as const
);
handyTypes.safe_integer = Object.assign(
  (v: unknown): v is number => Number.isSafeInteger(v),
  { category: "number" } as const
);

handyTypes.big_integer = Object.assign(
  (v: unknown): v is BigInt => typeof v === "bigint",
  { category: "bigint" } as const
);

handyTypes.positive_integer = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v > 0,
  { category: "number" } as const
);

handyTypes.negative_integer = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v < 0,
  { category: "number" } as const
);

handyTypes.non_positive_integer = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v < 1,
  { category: "number" } as const
);

handyTypes.non_negative_integer = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v >= 0,
  { category: "number" } as const
);

// byte wise integer
handyTypes["8bit_integer"] = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v >= -128 && v <= 127,
  { category: "number" } as const
);

handyTypes["8bit_unsigned_integer"] = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v >= 0 && v <= 255,
  { category: "number" } as const
);

handyTypes["16bit_integer"] = Object.assign(
  (v: unknown): v is number =>
    handyTypes.integer!(v) && v >= -32768 && v <= 32767,
  { category: "number" } as const
);

handyTypes["16bit_unsigned_integer"] = Object.assign(
  (v: unknown): v is number => handyTypes.integer!(v) && v >= 0 && v <= 65535,
  { category: "number" } as const
);

handyTypes["32bit_integer"] = Object.assign(
  (v: unknown): v is number =>
    handyTypes.integer!(v) && v >= -2147483648 && v <= 2147483647,
  { category: "number" } as const
);

handyTypes["32bit_unsigned_integer"] = Object.assign(
  (v: unknown): v is number =>
    handyTypes.integer!(v) && v >= 0 && v <= 4294967295,
  { category: "number" } as const
);

// String Types ----------------------------------------------------------------
handyTypes.string = Object.assign(
  (v: unknown): v is string => typeof v === "string",
  { category: "string" } as const
);
handyTypes.non_empty_string = Object.assign(
  (v: unknown): v is string => handyTypes.string!(v) && v !== "",
  { category: "string" } as const
);

handyTypes.trimmed_non_empty_string = Object.assign(
  (v: unknown): v is string => handyTypes.string!(v) && !!v.trim().length,
  { category: "string" } as const
);

// Object Types ----------------------------------------------------------------
handyTypes.object = Object.assign(
  (v: unknown): v is object | null => typeof v === "object",
  { category: "object" } as const
);
handyTypes.non_null_object = Object.assign(
  (v: unknown): v is object => v !== null && handyTypes.object!(v),
  { category: "object" } as const
);
handyTypes.plain_object = Object.assign(
  (v: unknown): v is object =>
    handyTypes.non_null_object!(v) && !Array.isArray(v),
  { category: "object" } as const
);

// Array Types -----------------------------------------------------------------
handyTypes.array = Object.assign(
  (v: unknown): v is unknown[] => Array.isArray(v),
  {
    category: "array",
  } as const
);
handyTypes.non_empty_array = Object.assign(
  (v: unknown): v is unknown[] => handyTypes.array!(v) && v.length > 0,
  { category: "array" } as const
);

handyTypes.function = Object.assign(
  (v: unknown): v is Function => typeof v === "function",
  { category: "function" } as const
);
handyTypes.symbol = Object.assign(
  (v: unknown): v is Symbol => typeof v === "symbol",
  { category: "symbol" } as const
);
handyTypes.boolean = Object.assign(
  (v: unknown): v is boolean => typeof v === "boolean",
  { category: "boolean" } as const
);

handyTypes.nan = Object.assign(
  (v: unknown): v is typeof NaN => Number.isNaN(v),
  {
    category: "other",
  } as const
);

// Other Types  ----------------------------------------------------------------
handyTypes.any = Object.assign((v: unknown): v is any => true, {
  category: "other",
} as const);
handyTypes.nullish = Object.assign(
  (v: unknown): v is NullishValue => v === undefined || v === null,
  { category: "other" } as const
);
handyTypes.non_nullish = Object.assign(
  (v: unknown): v is Exclude<AllBasicTypes, NullishValue> =>
    !handyTypes.nullish!(v),
  { category: "other" } as const
);

// ---------------------------- Type Names -------------------------------------
type TypeNames = Readonly<Record<keyof HandyTypes_Interface, string>>;

const typeNames: TypeNames = {
  // Base Types
  undefined: "Undefined",
  boolean: "Boolean",
  number: "Number",
  big_integer: "Big Integer",
  symbol: "Symbol",
  string: "String",
  object: "Object",
  function: "Function",

  finite_number: "Finite Number",

  positive_number: "Positive Number",
  non_negative_number: "Non Negative Number",

  negative_number: "Negative Number",
  non_positive_number: "Non Positive Number",

  integer: "Integer",
  safe_integer: "Safe Integer",

  positive_integer: "Positive Integer",
  non_negative_integer: "Non Negative Integer",

  negative_integer: "Negative Integer",
  non_positive_integer: "Non Positive Integer",

  "8bit_integer": "8 Bit Integer",
  "8bit_unsigned_integer": "8 Bit Unsigned Integer",
  "16bit_integer": "16 Bit Integer",
  "16bit_unsigned_integer": "16 Bit Unsigned Integer",
  "32bit_integer": "32 Bit Integer",
  "32bit_unsigned_integer": "32 Bit Unsigned Integer",

  non_empty_string: "Non-Empty String",
  trimmed_non_empty_string: "Non-Empty String (trimmed)",

  plain_object: "Plain Object",
  non_null_object: "Non-Null Object",

  array: "Array",
  non_empty_array: "Non-Empty Array",

  nan: "NaN",
  any: "Any",
  nullish: "Nullish",
  non_nullish: "Non-Nullish",
};

const frozenTypeObject = Object.freeze(
  handyTypes
) as Readonly<HandyTypes_Interface>;
const frozenTypeNames = Object.freeze(typeNames);

export type AllHandyTypes = keyof HandyTypes_Interface;

export { frozenTypeObject as handyTypes, frozenTypeNames as typeNames };
